power: stm32l5: support of the Low Power sleep mode

FRASTM opened the pull request:
With this patch, the low power mode on stm32L5 is enabling the Low Power Timer to measure the sleep time duration and wakeup from sleep mode on this LP timer expiration.
The SLEEP modes (corresponding to STM32 stop 0,1,2 modes) are supported on this device.
When low power is enabled on the system and when going to sleep, the lptim counts and wakes up the kernel though its interrupt routine.
The soc policy decides to go to sleep, if compatible with its residency when calling the sys_pm_policy_next_state for a nb of kernel ticks.

This PR is requires the support of the LPTimer on stm32l5  #32999

Note that DEEP_SLEEP modes are not enabled yet, as LPTIM cannot run during stm32 standby or shutdown modes.

Signed-off-by: Francois Ramu francois.ramu@st.com

FRASTM committed:
drivers: timer: st_stm32: add lptimer management to stm32l5 series

This patch introduces the support of the LowPower Timer
for the STM32L5xx from STMicroelectronics.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Merge lptim an PM is an single PR
Rebase
Set the LSE clock as default source clock for stm32l5xx boards. Consequently the LSE oscillator propagation for system clock must be  enabled for this soc serie. Setting the LSE system clock (LSESYS) bit of the RCC BDCR register.
Add a fix () to avoid multiple declaration of `z_clock_set_timeout` etc. when the STM32_LPTIM_TIMER is configured (not compatible with CORTEX_M_SYSTICK)

FRASTM commented:
Rebasing with the definition of the CORTEX_M_SYSTICK
LPTIM source clock are set to LSE

FRASTM committed:
dts: arm: stm32l5: add driver lptimer management to stm32l5 series

This patch introduces the support of the Low Power Timer
for the STM32L5xx from STMicroelectronics.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on soc/arm/st_stm32/stm32l5/power.c:
```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2019 STMicroelectronics.
```
2021

erwango commented on soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series:
```diff
@@ -10,4 +10,14 @@ source "soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l5*"
 
+if PM
+config PM_DEVICE
+	bool
```
Remove bool

FRASTM replied:
ok

erwango commented on soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series:
```diff
@@ -10,4 +10,14 @@ source "soc/arm/st_stm32/stm32l5/Kconfig.defconfig.stm32l5*"
+	default y
+
+config STM32_LPTIM_TIMER
+	bool
```
Remove bool

FRASTM replied:
ok

erwango commented on soc/arm/st_stm32/stm32l5/power.c:
```diff
@@ -0,0 +1,138 @@
+
+	ARG_UNUSED(dev);
+
+	ret = irq_lock();
```
I don't think this is required

FRASTM replied:
removed

erwango commented on soc/arm/st_stm32/stm32l5/power.c:
```diff
@@ -0,0 +1,138 @@
+		break;
+	case 2:
+		/* this corresponds to the STOP2 mode: */
+#ifdef CONFIG_DEBUG
+		/* Enable the Debug Module during STOP mode */
+		LL_DBGMCU_EnableDBGStopMode();
+#endif /* CONFIG_DEBUG */
```
I'm not sure we need to do this at this step.
So maybe do it once for all at init.

FRASTM replied:
yes, put in the `stm32_power_init() `
Note this is valid for all the low power modes : STOP 0/1/2 and to be completed with LL_DBGMCU_EnableDBGSleepMode if needed.

FRASTM committed:
soc: arm: stm32l5 is supporting PM with lptim

The low power timer of the soc is enabled when the PM is defined
for the STM32L5xx series.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
At this time, only **stm32l562e_dk** board has low power sleep modes in the stm32l5xx soc serie.

FRASTM commented:
rebase , fixing conflicts with stm32l562e_dk.dts

FRASTM committed:
boards: arm: stm32l562 disco kit enable MSI autocalibration

Enable MSI hardware auto calibration for this stm32l562
disco board. It depends on the STM_LSE clock.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
adding a commit to enable the MSI auto-calibration on the disco board

erwango commented on boards/arm/stm32l562e_dk/stm32l562e_dk.dts:
```diff
@@ -21,12 +21,37 @@
 		zephyr,flash = &flash0;
 	};
 
+	power-states {
+		stop0: state0 {
+			compatible = "zephyr,power-state";
+			power-state-name = "suspend-to-idle";
+			substate-id = <0>;
+			min-residency-us = <100>;
+		};
+		stop1: state1 {
+			compatible = "zephyr,power-state";
+			power-state-name = "suspend-to-idle";
+			substate-id = <1>;
+			min-residency-us = <500>;
+		};
+		stop2: state2 {
+			compatible = "zephyr,power-state";
+			power-state-name = "suspend-to-idle";
+			substate-id = <2>;
+			min-residency-us = <900>;
+		};
+	};
```
Please move this to `boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi`

FRASTM replied:
ok, moved to boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi

FRASTM committed:
soc: arm: stm32l5 serie with low power management

The low power modes are available on the stm32l5 soc
with the mcu STOP0/1/2 modes, depending on the CONFIG_PM

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: st_stm32: add lptimer to the stm32l562e dk board

This patch introduces the support of the LowPower Timer
for the STM32l562ze disco kit (sec. and non-sec. version)
LSE clock is selected as LPTIM clock source on this board.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32l562 disco kit has low power states

The low power states 0∕1∕2 are added to the stm32l562e_dk board
with default min residency values.
The substate-id is mapping the same Zephyr power state.
They correspond to the stm32l562 low power stop0/1/2 modes.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
- moving power state definition to boards/arm/stm32l562e_dk/stm32l562e_dk_common.dtsi
- adding the lptim1 and low power modes to the boards/arm/stm32l562e_dk_ns 
- change the substate-id range from 1-3 as described in the _include/power/power_state.h_
 this is changing the range in `pm_power_state_set() `of the soc/arm/st_stm32/stm32l5/power.c

ABOSTM commented on soc/arm/st_stm32/stm32l5/power.c:
```diff
@@ -0,0 +1,123 @@
+		switch (info.substate_id) {
+		case 0:	/* STOP0 */
+			__fallthrough;
+		case 1:	/* STOP1 */
```
Suported substate_id are 1, 2, 3.
substate_id = 1 correspond to STOP0 in dtsi file

FRASTM replied:
correct

ABOSTM reviewed (changes requested):
Commit Message:
```
The low power states 0∕1∕2 are added to the stm32l562e_dk board
with (default min residency values).
```
Sentence not finished: With what ? (or wrong parenthesis)

FRASTM commented:
rebase

erwango commented on boards/arm/stm32l562e_dk/stm32l562e_dk_defconfig:
```diff
@@ -21,6 +21,7 @@ CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
 # PLL configuration
 CONFIG_CLOCK_STM32_PLL_SRC_MSI=y
 CONFIG_CLOCK_STM32_MSI_RANGE=6
+CONFIG_CLOCK_STM32_MSI_PLL_MODE=y
```
This symbol has a dependency on CONFIG_CLOCK_STM32_LSE=y. It should be added

FRASTM replied:
could be 
```
config CLOCK_STM32_MSI_PLL_MODE
	default y
	depends on STM32_LPTIM_CLOCK_LSE
```
in the soc/arm/st_stm32/stm32l5/Kconfig.defconfig.series

ABOSTM reviewed (approved):
LGTM

FRASTM commented:
rebase
add the CONFIG_CLOCK_STM32_LSE  in the board def_config to enable the CONFIG_CLOCK_STM32_MSI_PLL_MODE without warning (as the CLOCK_STM32_MSI_PLL_MODE depends on the CLOCK_STM32_LSE through the Kconfig.stm32l4_l5_wb_wl)
